sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala

```diff
@@ -247,58 +259,137 @@ object CTESubstitution extends Rule[LogicalPlan] {
+          resolvedCTERelations.filterNot {
+            case (cteName, cteDef) => cteDef.conf.resolver(cteName, name)
+          }
+        } else {
+          resolvedCTERelations
+        }
+        traverseAndSubstituteCTE(relation, forceInline, nonConflictingCTERelations, cteDefs)._1
       }
-      // CTE definition can reference a previous one
-      val substituted = substituteCTE(innerCTEResolved, alwaysInline, resolvedCTERelations)
-      val cteRelation = CTERelationDef(substituted)
+
+      // If recursion is allowed then it has higher priority than outer or previous relations so
+      // construct a not yet substituted but recursive `CTERelationDef`, that we will prepend to
+      // `resolvedCTERelations`.
+      val recursiveCTERelation = if (allowRecursion) {
+        Some(name -> CTERelationDef(relation, recursive = true))
+      } else {
+        None
+      }
+      // CTE definition can reference a previous one or itself if recursion allowed.
+      val (substituted, recursionFound) = substituteCTE(innerCTEResolved, alwaysInline,
+        resolvedCTERelations, recursiveCTERelation)
+      val cteRelation = recursiveCTERelation
+        .map(_._2.copy(child = substituted, recursive = recursionFound))
+        .getOrElse(CTERelationDef(substituted))
       if (!alwaysInline) {
         cteDefs += cteRelation
       }
+
+      // From this point any reference to the definition is non-recursive.
+      val nonRecursiveCTERelation = if (cteRelation.recursive) {
+        cteRelation.copy(recursive = false)
+      } else {
+        cteRelation
+      }
+
       // Prepending new CTEs makes sure that those have higher priority over outer ones.
-      resolvedCTERelations +:= (name -> cteRelation)
+      resolvedCTERelations +:= (name -> nonRecursiveCTERelation)
     }
     resolvedCTERelations
   }
 
+  /**
+   * @return tuple of the logical plan and recursionFound boolean.
+   */
   private def resolveWithCTERelations(
       table: String,
       alwaysInline: Boolean,
       cteRelations: Seq[(String, CTERelationDef)],
-      unresolvedRelation: UnresolvedRelation): LogicalPlan = {
-    cteRelations
+      unresolvedRelation: UnresolvedRelation): (LogicalPlan, Boolean) = {
+    var recursionFound = false
+    val cteRelationsOut = cteRelations
       .find(r => conf.resolver(r._1, table))
       .map {
         case (_, d) =>
           if (alwaysInline) {
             d.child
           } else {
+            if (d.recursive) {
+              recursionFound = true
+            }
             // Add a `SubqueryAlias` for hint-resolving rules to match relation names.
-            SubqueryAlias(table, CTERelationRef(d.id, d.resolved, d.output, d.isStreaming))
+            SubqueryAlias(table,
+              CTERelationRef(d.id, d.resolved, d.output, d.isStreaming, recursive = d.recursive))
           }
       }
       .getOrElse(unresolvedRelation)
+    (cteRelationsOut, recursionFound)
```

Comment:
Do we really need to return this flag? The caller can get it from the returned `CTERelationRef`

Reply:
Looking at the whole PR, only one caller needs the `recursionFound`. I suggest we just return the logical plan and let that single caller figure out if recursion happens or not
```
val substituted = ...
val isRecursive = substituted.find {
  case ref: CTERelationRef => ref.recursive
}.isDefined
```

Reply:
Both things corrected now.